### PR TITLE
Adding allow_codec_switching parameter

### PR DESCRIPTION
**Pull request by sr1990**
This PR is related to https://github.com/google/shaka-packager/issues/542 . 
Adding allow_codec_switching=true, will allow switching between different codecs, if they have the same language, media type (audio, video etc) and container type.
```
$ packager \
  in=h264_baseline_360p_600.mp4,stream=audio,output=audio.mp4 \
  in=input_text.vtt,stream=text,output=output_text.vtt \
  in=h264_baseline_360p_600.mp4,stream=video,output=h264_360p.mp4 \
  in=h264_main_480p_1000.mp4,stream=video,output=h264_480p.mp4 \
  in=h264_main_720p_3000.mp4,stream=video,output=h264_720p.mp4 \
  in=h264_high_1080p_6000.mp4,stream=video,output=h264_1080p.mp4 \
  in=h265_high_1080p_6000.mp4,stream=video,output=h265_1080p.mp4 \ 
  --mpd_output h264.mpd --allow_codec_switching=true
```

### CODE REVIEW

File: packager/app/mpd_flags.cc [hunk 1]
```diff
@@ -62,3 +62,8 @@ DEFINE_bool(
     "completely."
     "Ignored if $Time$ is used in segment template, since $Time$ requires "
     "accurate Segment Timeline.");
+DEFINE_bool(allow_codec_switching,
```
Comment (kqyang) [inline packager/app/mpd_flags.cc]: This block is not formatted correctly. Please follow either the blocks from line 49 - 52 or the blocks from line 53 - 64. Alternatively, you may need to use clang-format: https://github.com/google/shaka-packager/blob/master/packager/tools/git/check_formatting.py

File: packager/app/mpd_flags.cc [hunk 2]
```diff
@@ -62,3 +62,8 @@ DEFINE_bool(
...
+            false,
+	    "if enabled, allow switching between different codecs, if they "
+	    "have the same language, media type (audio, video etc) and "
+	    "container type");
```
Comment (kqyang) [inline packager/app/mpd_flags.cc]: Can you add a period, i.e. "."  after type?

File: packager/mpd/base/adaptation_set.cc [hunk 3]
```diff
@@ -165,12 +165,14 @@ class RepresentationStateChangeListenerImpl
 
 AdaptationSet::AdaptationSet(const std::string& language,
                              const MpdOptions& mpd_options,
-                             uint32_t* counter)
+                             uint32_t* counter,
+                             const MediaInfo& media_info)
```
Comment (kqyang) [inline packager/mpd/base/adaptation_set.cc]: Move "const MediaInfo& media_info" before "uint32_t* counter".

We follow Google C++ style guide, where input parameters are always placed before output parameters: https://google.github.io/styleguide/cppguide.html#Output_Parameters.

I thought about it again. This change is probably not needed if we don't plan to refactor out ProtectedContentMap. 

Instead, what you need is to change line 85 of period.cc to include the bool flag.

File: packager/app/mpd_flags.cc (hunk 2 above)
Comment (kqyang) [reply packager/app/mpd_flags.cc]: s/switching/adaptive switching/

File: packager/mpd/base/adaptation_set.h [hunk 4]
```diff
@@ -135,6 +135,9 @@ class AdaptationSet {
   /// @param id is the new ID to be set.
   void set_id(uint32_t id) { id_ = id; }
 
+  /// Get media info.
+  const MediaInfo& getMediaInfo() const { return media_info_;}
```
Comment (kqyang) [inline packager/mpd/base/adaptation_set.h]: s/getMediaInfo/media_info/ to be consistent.

File: packager/mpd/base/mpd_utils.cc [hunk 5]
```diff
@@ -142,7 +142,7 @@ std::string GetBaseCodec(const MediaInfo& media_info) {
   return codec;
 }
 
-std::string GetAdaptationSetKey(const MediaInfo& media_info) {
+std::string GetAdaptationSetKey(const MediaInfo& media_info, bool no_codec) {
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc]: s/no_codec/ignore_codec/

File: packager/mpd/base/period.h [hunk 6]
```diff
@@ -130,7 +131,8 @@ class Period {
                const MediaInfo& media_info);
     // Check if the two adaptation sets are switchable.
     bool Switchable(const AdaptationSet& adaptation_set_a,
-                    const AdaptationSet& adaptation_set_b);
+                    const AdaptationSet& adaptation_set_b,
+                    const bool allow_codec_switching);
```
Comment (kqyang) [inline packager/mpd/base/period.h]: remove const here. We don't prefix primitive types with const in this code base.

File: packager/mpd/base/period.h [hunk 7]
```diff
@@ -88,7 +88,8 @@ class Period {
   virtual std::unique_ptr<AdaptationSet> NewAdaptationSet(
       const std::string& lang,
       const MpdOptions& options,
-      uint32_t* representation_counter);
+      uint32_t* representation_counter,
+      const MediaInfo& media_info);
```
Comment (kqyang) [inline packager/mpd/base/period.h]: Same as previous comment. Move media_info before representation_counter.

File: packager/mpd/base/period.cc [hunk 8]
```diff
@@ -112,12 +112,26 @@ AdaptationSet* Period::GetOrCreateAdaptationSet(
 
     for (AdaptationSet* adaptation_set : adaptation_sets) {
       if (protected_adaptation_set_map_.Switchable(*adaptation_set,
-                                                   *new_adaptation_set)) {
+                                                   *new_adaptation_set, false)) {
```
Comment (kqyang) [inline packager/mpd/base/period.cc]: The code here shouldn't need to be updated if you don't plan to refactor ProtectedAdaptationSetMap out.

Instead, what you need is to change line 85 to include the bool flag.

File: packager/mpd/base/mpd_utils.h [hunk 9]
```diff
@@ -47,7 +47,8 @@ std::string GetCodecs(const MediaInfo& media_info);
 std::string GetBaseCodec(const MediaInfo& media_info);
 
 // Returns a key made from the characteristics that separate AdaptationSets.
-std::string GetAdaptationSetKey(const MediaInfo& media_info);
+std::string GetAdaptationSetKey(const MediaInfo& media_info,
+           bool no_codec = false);
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.h]: We don't normally allow default parameters, which could hide problems. Please do not use default parameter here.

File: packager/mpd/base/adaptation_set.cc (hunk 3 above)
Comment (sr1990) [reply packager/mpd/base/adaptation_set.cc]: @kqyang Just including the bool flag on line 85 will not create a new adaptation set along with switching supplemental property. I will take a look into refactoring out ProtectedContentMap. Please let me know if you have any pointers.

Comment (kqyang) [reply packager/mpd/base/adaptation_set.cc]: > Just including the bool flag on line 85 will not create a new adaptation set along with switching supplemental property

There are a few additional changes associated with that. See below:

```c++
  // AdaptationSets with the same key could differ in ContentProtection or codec (depends
  // on whether codec switching is allowed).
  const std::string key = GetAdaptationSetKey(media_info, mpd_options_.allow_codec_switching);
  std::list<AdaptationSet*>& adaptation_sets = adaptation_set_list_map_[key];

  for (AdaptationSet* adaptation_set : adaptation_sets) {
    // TODO: Change Match to check if codecs is the same too.
    if (protected_adaptation_set_map_.Match(*adaptation_set, media_info))
      return adaptation_set;
  }

  // None of the adaptation sets match with the new content protection.
  // Need a new one.
  const std::string language = GetLanguage(media_info);
  std::unique_ptr<AdaptationSet> new_adaptation_set =
      NewAdaptationSet(language, mpd_options_, representation_counter_);
  if (!SetNewAdaptationSetAttributes(language, media_info, adaptation_sets,
                                     new_adaptation_set.get())) {
    return nullptr;
  }

  if (content_protection_in_adaptation_set &&
      media_info.has_protected_content()) {
    protected_adaptation_set_map_.Register(*new_adaptation_set, media_info);
    AddContentProtectionElements(media_info, new_adaptation_set.get());
  }

  for (AdaptationSet* adaptation_set : adaptation_sets) {
    if (protected_adaptation_set_map_.Switchable(*adaptation_set,
                                                 *new_adaptation_set)) {
      adaptation_set->AddAdaptationSetSwitching(new_adaptation_set.get());
      new_adaptation_set->AddAdaptationSetSwitching(adaptation_set);
    }
  }

```

> I will take a look into refactoring out ProtectedContentMap. Please let me know if you have any pointers.

That will be great. I think we can start with the change above and we can file a separate issue for the refactoring.

Comment (sr1990) [reply packager/mpd/base/adaptation_set.cc]: For comparing the codec in Match(),  do you think adding codec as AdaptationSet member and setting it via SetNewAdaptationSetAttributes make sense or is there any other way to access codec info?

Comment (kqyang) [reply packager/mpd/base/adaptation_set.cc]: I just realized that you can store the codec in AdaptationSet::UpdateFromMediaInfo function.

File: packager/mpd/base/adaptation_set.h (hunk 4 above)
Comment (sr1990) [reply packager/mpd/base/adaptation_set.h]: removed mediainfo from adaptation set

**Comment by sr1990** [comment]
@kqyang CI jobs are failing. Please review and let me know.

File: packager/mpd/base/period.cc [hunk 10]
```diff
@@ -240,11 +239,12 @@ const AdaptationSet* Period::FindOriginalAdaptationSetForTrickPlay(
   MediaInfo media_info_no_trickplay = media_info;
   media_info_no_trickplay.mutable_video_info()->clear_playback_rate();
 
-  std::string key = GetAdaptationSetKey(media_info_no_trickplay);
+  std::string key = GetAdaptationSetKey(media_info_no_trickplay, false);
```
Comment (kqyang) [inline packager/mpd/base/period.cc]: s/false/mpd_options_.mpd_params.allow_codec_switching/

It has to be the same as the key we use when setting up |adaptation_set_list_map_|, otherwise |adaptation_sets| below would be empty.

File: packager/mpd/base/period.cc [hunk 11]
```diff
@@ -240,11 +239,12 @@ const AdaptationSet* Period::FindOriginalAdaptationSetForTrickPlay(
...
   const std::list<AdaptationSet*>& adaptation_sets =
       adaptation_set_list_map_[key];
   for (AdaptationSet* adaptation_set : adaptation_sets) {
-    if (protected_adaptation_set_map_.Match(*adaptation_set, media_info)) {
+    if (protected_adaptation_set_map_.Match(*adaptation_set, media_info,
+                                            true)) {
```
Comment (kqyang) [inline packager/mpd/base/period.cc]: Looks like this is a bug in the original code. The flag content_protection_in_adaptation_set should have the same value as the one in line 93.

Can you pass content_protection_in_adaptation_set to SetNewAdaptationSetAttributes then to  FindOriginalAdaptationSetForTrickPlay?

File: packager/mpd/base/adaptation_set.cc [hunk 12]
```diff
@@ -412,10 +412,16 @@ void AdaptationSet::UpdateFromMediaInfo(const MediaInfo& media_info) {
```
Comment (kqyang) [inline packager/mpd/base/adaptation_set.cc]: Prefer just set 

codec_ = GetCodecs(media_info);

File: packager/mpd/base/period.cc [hunk 13]
```diff
@@ -260,7 +260,38 @@ void Period::ProtectedAdaptationSetMap::Register(
...
+        return false;
+
+    } else if (media_info.has_text_info() &&
+               media_info.text_info().has_codec()) {
+      if (adaptation_set_codec.compare(base::SplitString(
+              media_info.text_info().codec(), ".", base::TRIM_WHITESPACE,
+              base::SPLIT_WANT_NONEMPTY)[0]) != 0)
+        return false;
+    }
+  }
```
Comment (kqyang) [inline packager/mpd/base/period.cc]: Is SplitString needed? 

with the change suggested above in line 412 of adaptation_set.cc, we can just do:

if (adaptation_set.codec() != GetCodecs(media_info))
  return false;

File: packager/mpd/base/adaptation_set.h [hunk 14]
```diff
@@ -269,6 +269,9 @@ class AdaptationSet {
   // Determined by examining the MediaInfo passed to AddRepresentation().
   std::string content_type_;
 
+  // Codec of AdaptationSet
+  std::string codec_;
```
Comment (kqyang) [inline packager/mpd/base/adaptation_set.h]: Also define an accessor for codec_ after line 178:

/// @return codec.
std::string codec() const { return codec_; }

File: packager/app/test/packager_test.py [hunk 15]
```diff
@@ -1499,6 +1503,17 @@ def testLiveStaticProfileWithTimeInSegmentName(self):
...
+    streams = [
+        self._GetStream('video', test_file='bear-640x360-hevc.mp4'),
+        self._GetStream('video', test_file='bear-640x360.mp4'),
+        self._GetStream('video', test_file='bear-1280x720.mp4'),
+        self._GetStream('audio', test_file='bear-640x360.mp4'),
+    ]
+
+    self.assertPackageSuccess(streams, self._GetFlags(output_dash=True,allow_codec_switching=True))
+    self._CheckTestResults('audio-video-with-codec-switching')
+
```
Comment (kqyang) [inline packager/app/test/packager_test.py]: Can you add another variant with encryption and trick play?

```python
  def testAllowCodecSwitchingWithEncryptionAndTrickplay(self):
    streams = [
        self._GetStream('video', test_file='bear-640x360-hevc.mp4'),
        self._GetStream('video', test_file='bear-640x360.mp4'),
        self._GetStream('video', test_file='bear-1280x720.mp4'),
        self._GetStream('video', test_file='bear-1280x720.mp4', trick_play_factor=1),
        self._GetStream('audio', test_file='bear-640x360.mp4'),
    ]

    self.assertPackageSuccess(streams,
                              self._GetFlags(output_dash=True, 
                                             allow_codec_switching=True, 
                                             encryption=True))
    # Mpd cannot be validated right now since we don't generate determinstic
    # mpd with multiple inputs due to thread racing.
    # TODO(b/73349711): Generate determinstic mpd or at least validate mpd
    #                   schema.
    # See also https://github.com/google/shaka-packager/issues/177.
```

File: packager/app/test/packager_test.py [hunk 16]
```diff
@@ -1499,6 +1503,17 @@ def testLiveStaticProfileWithTimeInSegmentName(self):
...
+  def testAllowCodecSwitching(self):
+    streams = [
+        self._GetStream('video', test_file='bear-640x360-hevc.mp4'),
+        self._GetStream('video', test_file='bear-640x360.mp4'),
+        self._GetStream('video', test_file='bear-1280x720.mp4'),
+        self._GetStream('audio', test_file='bear-640x360.mp4'),
+    ]
+
+    self.assertPackageSuccess(streams, self._GetFlags(output_dash=True,allow_codec_switching=True))
+    self._CheckTestResults('audio-video-with-codec-switching')
```
Comment (kqyang) [inline packager/app/test/packager_test.py]: Unfortunately this line has to be removed as we don't generate determinstic mpd with multiple inputs right now.

Replace it with:

    # Mpd cannot be validated right now since we don't generate determinstic
    # mpd with multiple inputs due to thread racing.
    # TODO(b/73349711): Generate determinstic mpd or at least validate mpd
    #                   schema.
    # See also https://github.com/google/shaka-packager/issues/177.

File: packager/mpd/base/period.cc (hunk 13 above)
Comment (sr1990) [reply packager/mpd/base/period.cc]: GetCodecs returns the entire codec string (avc1.4d400d / avc1.4d401f) if container type is not webm. Should I change the GetCodecs to return media_info.video_info().codec().substr(0, 4); for all container types or keep the SplitString in Match?

Comment (kqyang) [reply packager/mpd/base/period.cc]: Ah, good catch, we shouldn't be using the entire codec string as two different representations in the same adaptation set could have two different codec strings.

Please replace GetCodecs with GetBaseCodec instead.

Comment (sr1990) [reply packager/mpd/base/period.cc]: TrickPlayWithMatchingAdaptationSet unit test is failing as the codec in default_adaptation_set_ is no call to UpdateFromMediaInfo.  
So, instead of setting codec from AdaptationSet::UpdateFromMediaInfo do you think it would make sense to set it at Period::SetNewAdaptationSetAttributes. If not, how do you change the unit test? I tried calling the UpdateFromMediaInfo directly but that did not work.

File: packager/mpd/base/period.cc (hunk 11 above)
Comment (sr1990) [reply packager/mpd/base/period.cc]: Yes.

File: packager/mpd/base/period.cc (hunk 13 above)
Comment (kqyang) [reply packager/mpd/base/period.cc]: Sure. You may add a AdaptationSet::set_codec(const std::string& codec) function and call it in Period::SetNewAdaptationSetAttributes. 

(This tells me that we should really refactor the code later to clean up the code and remove ProtectedContentMap)

**Comment by sr1990** [comment]
@googlebot I consent.

File: packager/app/test/packager_test.py [hunk 17]
```diff
@@ -1499,6 +1503,42 @@ def testLiveStaticProfileWithTimeInSegmentName(self):
...
+    #                   schema.
+    # See also https://github.com/google/shaka-packager/issues/177.
+
+  def testAllowCodecSwitchingWithEncryptionAndTrickplay(self):
+    streams = [
+        self._GetStream('video', test_file='bear-640x360-hevc.mp4'),
+        self._GetStream('video', test_file='bear-640x360.mp4'),
+        self._GetStream('video', test_file='bear-1280x720.mp4'),
+        self._GetStream('video', test_file='bear-1280x720.mp4',
+                                           trick_play_factor=1),
```
Comment (kqyang) [inline packager/app/test/packager_test.py]: I think trick_play_factor should align with 'video' above. You may use 'pyformat' to align it properly (there could be other formatting issues in this file, so it is fine to leave it as is for now and do it in a separate pull request or leave it to us to address).

File: packager/mpd/base/adaptation_set.h [hunk 18]
```diff
@@ -176,6 +176,13 @@ class AdaptationSet {
   /// @return true if it is a video AdaptationSet.
   bool IsVideo() const;
 
+  /// @return codec.
+  std::string codec() const { return codec_; }
```
Comment (kqyang) [inline packager/mpd/base/adaptation_set.h]: Use const reference to avoid unnecessary copying:

const std::string& codec() const { return codec_; }

File: packager/mpd/base/period.cc [hunk 19]
```diff
@@ -7,6 +7,7 @@
 #include "packager/mpd/base/period.h"
 
 #include "packager/base/stl_util.h"
+#include "packager/base/strings/string_split.h"
```
Comment (kqyang) [inline packager/mpd/base/period.cc]: Is this still needed?

File: packager/mpd/base/period.cc [hunk 20]
```diff
@@ -176,7 +173,8 @@ bool Period::SetNewAdaptationSetAttributes(
     const std::string& language,
     const MediaInfo& media_info,
     const std::list<AdaptationSet*>& adaptation_sets,
-    AdaptationSet* new_adaptation_set) {
+    AdaptationSet* new_adaptation_set,
+    bool content_protection_in_adaptation_set) {
```
Comment (kqyang) [inline packager/mpd/base/period.cc]: Swap the order of the last two parameters as Google C++ style guide requires the input argument to be placed before the output argument: https://google.github.io/styleguide/cppguide.html#Output_Parameters

File: packager/mpd/base/period.cc (hunk 19 above)
Comment (sr1990) [reply packager/mpd/base/period.cc]: no. removed it.

File: packager/app/test/packager_test.py (hunk 15 above)
Comment (kqyang) [reply packager/app/test/packager_test.py]: Ooops, I forgot one additional line here in my previous comment:

```python
    self._CheckTestResults(
        'audio-video-with-codec-switching-encryption-trick-play',
        diff_files_policy=DiffFilesPolicy(
            allowed_diff_files=['output.mpd'], exact=False))
```

Similar for the test above:

```
    self._CheckTestResults(
        'audio-video-with-codec-switching',
        diff_files_policy=DiffFilesPolicy(
            allowed_diff_files=['output.mpd'], exact=False))
```

File: packager/mpd/public/mpd_params.h [hunk 21]
```diff
@@ -80,6 +80,9 @@ struct MpdParams {
   /// SegmentTimeline if it is enabled. It will be populated from segment
   /// duration specified in ChunkingParams if not specified.
   double target_segment_duration = 0;
+  /// If enabled, allow switching between different codecs, if they have the
+  /// same language, media type (audio, video etc) and container type.
+  bool allow_codec_switching = false;
```
Comment (kqyang) [inline packager/mpd/public/mpd_params.h]: Normally we don't pull changes from master, which may cause confusions to the bots as seen in this PR. I'll see if I can merge it with this later.

**Comment by sr1990** [comment]
@kqyang Do you think it makes sense to check and add audio channel count to adaptation set key in mpd_utils.cc:GetAdaptationSetKey when allow_codec_switching parameter is set to true ?

**Comment by kqyang** [comment]
Audio streams with different number of channels can be put in the same AdaptationSet, regardless of the `allow_codec_switching` flag, so I don't think it should be included in `GetAdaptationSetKey`. Let me know if you have a different opinion.

**Comment by sr1990** [comment]
I think adaptive switch between audio with different channels won't be seamless. On Exoplayer side, there is a flag allowAudioMixedChannelCountAdaptiveness in DefaultTrackSelector which is set to false by default and it filters out the audio track with appropriate channel count at: https://github.com/google/ExoPlayer/blob/e822196336b64208ba7caa557e4c6ee02b03335e/library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java#L1647 and 
as per issue https://github.com/google/ExoPlayer/issues/6497: 
`"Also, by default, we won't allow adaptation between different channel counts (i.e. stereo and 5.1) because there may be an audible discontinuity when we switch formats."`
This flag has been added in 2.10.5 as per https://github.com/google/ExoPlayer/blob/release-v2/RELEASENOTES.md. 
So for clients depending on older versions of exoplayer and other players that do not handle this case, we could add a flag (--allow_audio_mixed_channel_count set to true by default) on the packager side to disable audio representations with different channel counts in the same adaptation set. What do you think?

**Comment by sr1990** [comment]
I think adaptive switch between audio with different channels won't be seamless.
On Exoplayer side, there is a flag allowAudioMixedChannelCountAdaptiveness in DefaultTrackSelector which is set to false by default and it filters out the audio track with appropriate channel count at: https://github.com/google/ExoPlayer/blob/e822196336b64208ba7caa557e4c6ee02b03335e/library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java#L1647 and
as per issue https://github.com/google/ExoPlayer/issues/6497:
`"Also, by default, we won't allow adaptation between different channel counts (i.e. stereo and 5.1) because there may be an audible discontinuity when we switch formats."`
This flag has been added in 2.10.5 as per https://github.com/google/ExoPlayer/blob/release-v2/RELEASENOTES.md.

So for players that do not handle this case, we could add a flag (--allow_audio_mixed_channel_count set to true by default) on the packager side to disable audio representations with different channel counts in the same adaptation set or we could let the players handle it. What do you think?

**Comment by kqyang** [comment]
@sr1990 So ExoPlayer already handles the case of a single AdaptationSet with audio streams of different number of channels, the feature for packager is not helpful for ExoPlayer. 

It seems that it may create problems instead, as people would have to package two sets of contents if they want to enable adaptive switching between different audio streams in one platform but disable it in another platform.

**Comment by sr1990** [comment]
`It seems that it may create problems instead, as people would have to package two sets of contents if they want to enable adaptive switching between different audio streams in one platform but disable it in another platform`
I agree. Thanks.
